pkg/nuctl/command/common/importreport.go

```diff
@@ -0,0 +1,138 @@
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/nuclio/logger"
+)
+
+type ProjectReports struct {
+	projectReports map[string]*ProjectReport
+}
+
+func NewProjectsReport() *ProjectReports {
+	return &ProjectReports{
+		projectReports: make(map[string]*ProjectReport),
+	}
+}
+
+func (psr *ProjectReports) AddReport(report *ProjectReport) {
+	psr.projectReports[report.Name] = report
+}
+
+func (psr *ProjectReports) GetOrCreate(projectName string) *ProjectReport {
+	report, found := psr.projectReports[projectName]
+	if !found {
+		report = NewProjectReport(projectName)
+		psr.AddReport(report)
+	}
+	return report
+}
+
+func (psr *ProjectReports) SaveToFile(ctx context.Context, loggerInstance logger.Logger, path string) {
+	saveReportToFile(ctx, loggerInstance, psr.projectReports, path)
+}
+
+func (psr *ProjectReports) SprintfError() string {
+	report := ""
+	for _, projectReport := range psr.projectReports {
+		report += projectReport.SprintfError()
+	}
+	return report
+}
+
+type ProjectReport struct {
+	Name            string
+	Skipped         bool
+	Success         bool
+	Failed          *FailReport
+	FunctionReports *FunctionReports
+}
+
+func NewProjectReport(name string) *ProjectReport {
+	return &ProjectReport{
+		Name:            name,
+		FunctionReports: NewFunctionReports(),
+	}
+}
+
+// SetFailed sets the "Failed" status if it hasn't been set yet. In other words, it retains only the first error that occurred.
+func (pr *ProjectReport) SetFailed(failReport *FailReport) {
+	if pr.Failed == nil {
+		pr.Failed = failReport
+	}
+}
+
+func (pr *ProjectReport) SaveToFile(ctx context.Context, loggerInstance logger.Logger, path string) {
+	saveReportToFile(ctx, loggerInstance, pr.FunctionReports, path)
+}
+
+func (pr *ProjectReport) SprintfError() string {
+	if pr.Failed != nil {
+		return fmt.Sprintf("Failed to import project `%s`. Reason: %s", pr.Name, pr.Failed.FailReason) + pr.FunctionReports.SprintfError()
+	}
+	return ""
+}
+
+type FunctionReports struct {
+	Success []string
+	Failed  map[string]*FailReport
+
+	mutex sync.Mutex
+}
+
+func NewFunctionReports() *FunctionReports {
+	return &FunctionReports{Failed: make(map[string]*FailReport), mutex: sync.Mutex{}}
+}
+
+func (frs *FunctionReports) SaveToFile(ctx context.Context, loggerInstance logger.Logger, path string) {
+	saveReportToFile(ctx, loggerInstance, frs, path)
+}
+
+func (frs *FunctionReports) SprintfError() string {
+	report := ""
+	for name, failReason := range frs.Failed {
+		report += fmt.Sprintf("Failed to import function `%s`. Reason: %s.", name, failReason.FailReason)
+	}
+
+	return report
+}
+
+func (frs *FunctionReports) AddFailure(name string, err error) {
+	frs.mutex.Lock()
+	defer frs.mutex.Unlock()
+
+	frs.Failed[name] = &FailReport{
+		FailReason: err.Error(),
+	}
+}
+
+func (frs *FunctionReports) AddSuccess(name string) {
+	frs.mutex.Lock()
+	defer frs.mutex.Unlock()
+
+	frs.Success = append(frs.Success, name)
+}
+
+type FailReport struct {
```

Comment:
This is very similar to the patchManifest's `failDescription`. Perhaps they can be merged and used in both places (use `retryable` will be used as "Autofixable") .

Reply:
Resolving as it was decided not to merge them due to insufficient commonalities.